### DISCUSSION TITLE
Custom render loop and integration with other engines

### DISCUSSION THREAD

**Discussion by damix911** [pixijs/pixijs#7210] [Q&A]
Hi everybody, thank you for your work on this awesome library.

I need to integrate PixiJS with another, custom engine, that owns its own render loop.

**Question 1.** _Why does this code does not display anything? See on [CodePen](https://codepen.io/dawken/pen/jOVNJad?editors=1010)_

```js
const renderer = new PIXI.Renderer({ context: document.getElementById("canvas").getContext("webgl") });
const stage = new PIXI.Container();

const ticker = new PIXI.Ticker();
ticker.add(() => {
  renderer.render(stage);
}, PIXI.UPDATE_PRIORITY.LOW);
ticker.start();

const sprite = PIXI.Sprite.from(bird);
sprite.anchor.set(0.5);
sprite.position.set(320, 180);

stage.addChild(sprite);
```

If I change the first line to:

```js
const renderer = new PIXI.Renderer({ view: document.getElementById("canvas") });
```

It works: the canvas gets cleared to black and the sprite is rendered; but:

**Question 2.** _Why does it resize the canvas?_ I would not want PixiJS to mess with canvas size, because it's owned by the other engine.

Also:

**Question 3.** _If I delete `ticker.start()`the sprite does not get rendered, but it looks like the canvas is still cleared to black; why?_ Note that I would not want PixiJS to issue clear commands, nor at render creation time nor at render time, because I must draw on top of other content rendered with the other engine. The docs for `Renderer` say something about `preserveDrawingBuffer` but I cannot do it because the context is pre-existing and owned by the other engine.

Finally:

**Question 4.** _Assuming that I get that micro sample to work properly, I will need to integrate the existing custom WebGL engine. When do I need to call `renderer.reset()`? I have been struggling with that; I tend to get `[.WebGL-0x7fd105545e00]RENDER WARNING: there is no texture bound to the unit ...` errors. See also [this question on StackOverflow](https://stackoverflow.com/questions/65913943/telling-pixijs-that-the-webgl-state-has-been-modified-externally).

Thanks! Let me know if I can help!

**Comment by ivanpopelyshev** [comment]
No way you'll get answer on stackoverflow about that.

Answer 1. because you have to pass the `view` too

Answer 2. Debug, patch the appropriate function. PixiJS is not a black box. You can do it from outside like `Renderer.prototype.something=function() { ... }`

Answer 3. honestly, I dont know, debug it 

Answer 4. PixiJS takes care of its textures, but it cant restore whatever your custom webgl engine state was. `reset()` should exist both on pixi side and on your side. For me, it worked with ThreeJS and BabylonJS just fine.

**Reply by damix911** [reply]
Thank you, I'll do more debugging. There is indeed a reset on both ends, so I am confident that I should be able to make it work.

Thanks!
